File: Csharp/SolidCourse/LSP/After.cs
// Base class
public abstract class MyBird
{
    // We won't have a Fly method here anymore.
}

// New Flyable interface
public interface IFlyable
{
    void Fly();
}

// Derived class 1 - Eagle
public class MyEagle : Bird, IFlyable
{
    public void Fly()
    {
        Console.WriteLine("Sparrow flying high!");
    }
}

// Derived class 2 - Penguin
public class MyPenguin : MyBird
{
    // Penguins have no flying capabilities and do not implement IFlyable
}

// Client Code
public class MyBirdManager
{
    public void MakeBirdFly(IFlyable bird)
    {
        bird.Fly(); // This works for Eagle and does not apply to Penguin.
    }
}


// var birdManager = new MyBirdManager();
// var eagle = new MySparrow();
// birdManager.MakeBirdFly(eagle);

File: CPlusPlus/LSP/After.cpp
#include <iostream>
#include <memory>

// Abstract base class for birds
class MyBird {
public:
    virtual ~MyBird() = default;
};

// Interface-like abstract class for flying behavior
class IFlyable {
public:
    virtual void Fly() = 0;
    virtual ~IFlyable() = default;
};

// Eagle (flyable)
class MyEagle : public MyBird, public IFlyable {
public:
    void Fly() override {
        std::cout << "Sparrow flying high!" << std::endl;
    }
};

// Penguin (non-flyable)
class MyPenguin : public MyBird {
    // Does not implement IFlyable
};

// Bird manager using interface
class MyBirdManager {
public:
    void MakeBirdFly(IFlyable& bird) {
        bird.Fly(); // Safe: Only works on flyable birds
    }
};
